File: 1872-stone-game-viii/1872-stone-game-viii.cpp
class Solution {
public:
    int stoneGameVIII(vector<int>& stones) {
        int n = stones.size();
        vector<int> pre(n+1, 0);
        for(int i = 0; i<n; ++i) pre[i+1] = pre[i] + stones[i];
        vector<int> alice(n+1, 0), bob(n+1, 0);
        alice[n] = pre[n], bob[n] = -pre[n];
        for(int i = n-1; i>=2; --i){
            alice[i] = max(alice[i+1], pre[i]+bob[i+1]);
            bob[i] = min(bob[i+1], -pre[i]+alice[i+1]);
        }
        return alice[2];
    }
};

File: 1872-stone-game-viii/1872-stone-game-viii.java
class Solution {
    public int stoneGameVIII(int[] stones) {
        int n = stones.length;
        int[] pre = new int[n+1];

        for(int i = 0; i<n; ++i) pre[i+1] = pre[i]+stones[i];

        int[] alice = new int[n+1];
        int[] bob = new int[n+1];
        alice[n] = pre[n];
        bob[n] = -pre[n];

        for(int i = n-1; i>=2; --i){
            alice[i] = Math.max(alice[i+1], pre[i]+bob[i+1]);
            bob[i] = Math.min(bob[i+1], -pre[i]+alice[i+1]);
        }
        return alice[2];
    }
}